File: dsa-problems/AdvancedGraphs/solutions/cpp/AGR-007-eulerian-trail-directed.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <deque>

using namespace std;

class Solution {
    vector<vector<int>> adj;
    vector<int> trail;

    void dfs(int u) {
        while (!adj[u].empty()) {
            int v = adj[u].back();
            adj[u].pop_back();
            dfs(v);
        }
        trail.push_back(u);
    }

public:
    vector<int> eulerTrail(int n, const vector<pair<int, int>>& edges) {
        int m = edges.size();
        if (m == 0) return {0};

        vector<int> in(n, 0), out(n, 0);
        adj.assign(n, vector<int>());

        for (const auto& e : edges) {
            out[e.first]++;
            in[e.second]++;
            adj[e.first].push_back(e.second);
        }

        int startNode = -1;
        int endNode = -1;
        int diffCount = 0;

        for (int i = 0; i < n; i++) {
            if (out[i] == in[i] + 1) {
                if (startNode != -1) return {};
                startNode = i;
                diffCount++;
            } else if (in[i] == out[i] + 1) {
                if (endNode != -1) return {};
                endNode = i;
                diffCount++;
            } else if (in[i] != out[i]) {
                return {};
            }
        }

        if (diffCount == 0) {
            for (int i = 0; i < n; i++) {
                if (out[i] > 0) {
                    startNode = i;
                    break;
                }
            }
        } else if (diffCount != 2) {
            return {};
        }

        if (startNode == -1) return {};

        dfs(startNode);

        if (trail.size() != m + 1) return {};

        reverse(trail.begin(), trail.end());
        return trail;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;
    vector<pair<int, int>> edges(m);
    for (int i = 0; i < m; i++) {
        cin >> edges[i].first >> edges[i].second;
    }

    Solution solution;
    vector<int> trail = solution.eulerTrail(n, edges);
    if (trail.empty()) {
        cout << "NO";
    } else {
        cout << "YES\n";
        for (int i = 0; i < (int)trail.size(); i++) {
            if (i) cout << ' ';
            cout << trail[i];
        }
    }
    return 0;
}

File: dsa-problems/AdvancedGraphs/solutions/java/AGR-007-eulerian-trail-directed.java
import java.util.*;

class Solution {
    private List<Deque<Integer>> adj;
    private List<Integer> trail;

    public int[] eulerTrail(int n, int[][] edges) {
        int m = edges.length;
        if (m == 0) return new int[]{0};

        int[] in = new int[n];
        int[] out = new int[n];
        adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayDeque<>());

        for (int[] e : edges) {
            out[e[0]]++;
            in[e[1]]++;
            adj.get(e[0]).add(e[1]);
        }

        int startNode = -1;
        int endNode = -1;
        int diffCount = 0;

        for (int i = 0; i < n; i++) {
            if (out[i] == in[i] + 1) {
                if (startNode != -1) return null; // More than one start
                startNode = i;
                diffCount++;
            } else if (in[i] == out[i] + 1) {
                if (endNode != -1) return null; // More than one end
                endNode = i;
                diffCount++;
            } else if (in[i] != out[i]) {
                return null; // Invalid degree
            }
        }

        if (diffCount == 0) {
            // Circuit: find first node with edges
            for (int i = 0; i < n; i++) {
                if (out[i] > 0) {
                    startNode = i;
                    break;
                }
            }
        } else if (diffCount != 2) {
            return null;
        }

        if (startNode == -1) return null; // Should not happen if m > 0

        trail = new ArrayList<>();
        dfs(startNode);

        if (trail.size() != m + 1) return null; // Disconnected components

        int[] res = new int[trail.size()];
        for (int i = 0; i < trail.size(); i++) {
            res[i] = trail.get(trail.size() - 1 - i);
        }
        return res;
    }

    private void dfs(int u) {
        Deque<Integer> neighbors = adj.get(u);
        while (!neighbors.isEmpty()) {
            int v = neighbors.pollLast();
            dfs(v);
        }
        trail.add(u);
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[][] edges = new int[m][2];
        for (int i = 0; i < m; i++) {
            edges[i][0] = sc.nextInt();
            edges[i][1] = sc.nextInt();
        }

        Solution solution = new Solution();
        int[] trail = solution.eulerTrail(n, edges);
        if (trail == null) {
            System.out.print("NO");
        } else {
            StringBuilder sb = new StringBuilder();
            sb.append("YES\n");
            for (int i = 0; i < trail.length; i++) {
                if (i > 0) sb.append(' ');
                sb.append(trail[i]);
            }
            System.out.print(sb.toString());
        }
        sc.close();
    }
}
